nomad/rpc.go

```diff
@@ -68,26 +68,39 @@ func NewServerCodec(conn io.ReadWriteCloser) rpc.ServerCodec {
 }
 
 // listen is used to listen for incoming RPC connections
-func (s *Server) listen() {
+func (s *Server) listen(ctx context.Context) {
 	for {
+		select {
+		case <-ctx.Done():
+			s.logger.Println("[INFO] nomad.rpc: Closing server RPC connection")
+			return
+		default:
+		}
+
 		// Accept a connection
 		conn, err := s.rpcListener.Accept()
 		if err != nil {
 			if s.shutdown {
 				return
 			}
+
+			select {
+			case <-ctx.Done():
+				return
```

Comment:
Need an empty default case here to prevent blocking until ctx.Done() is closed:

```go
select {
  case <-ctx.Done():
    return
  default:
}
```

Reply:
Good catch. See here: https://github.com/hashicorp/nomad/pull/3492/files#diff-7e792a308b759af800f2cd08ffe94c5cR90